### PR TITLE
Revert "Pull out check* from expression.Expression + move them to expressionsem + make most of them private (#15775)"

### PR DESCRIPTION
**Pull request by ibuclaw**
This reverts commit 9266f92fe5d3499c9f7345978aacd3775c1a21a3.

### CODE REVIEW

**Review by thewilsonator — DISMISSED** [review]
You've put a GDC tag on it this time, which is better than nothing, but I'd still like an explanation as to why this is necessary, considering that this undies progress of moving stuff out of AST and into sema. Again I suspect that the appropriate course of action here is to make this work not to  revert it.

_Please do not merge this without a satisfactory explanation_.

**Comment by ibuclaw** [comment]
```
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable4growMFNaNbZv':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:310: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toStringMxFNaNbNiNjZAxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable10allocValueMFNaNbMAxaC3dmd5mtype4TypeZk':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:268: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:269: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable4growMFNaNbZv':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:310: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toStringMxFNaNbNiNjZAxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable10allocValueMFNaNbMAxaC3dmd5mtype4TypeZk':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:268: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:269: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable4growMFNaNbZv':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:310: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toStringMxFNaNbNiNjZAxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable10allocValueMFNaNbMAxaC3dmd5mtype4TypeZk':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:268: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:269: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable4growMFNaNbZv':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:310: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toStringMxFNaNbNiNjZAxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable10allocValueMFNaNbMAxaC3dmd5mtype4TypeZk':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:268: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:269: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable4growMFNaNbZv':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:310: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toStringMxFNaNbNiNjZAxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/func.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable10allocValueMFNaNbMAxaC3dmd5mtype4TypeZk':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:268: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: /build/gcc-trunk/gcc/d/dmd/root/stringtable.d:269: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue7lstringMFNaNbNiNjZPa'
/usr/bin/ld: d/mtype.o: in function `_D3dmd4root11stringtable34__T11StringTableTC3dmd5mtype4TypeZ11StringTable8findSlotMxFNaNbNimMAxaZm':
/build/gcc-trunk/gcc/d/dmd/root/stringtable.d:292: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
/usr/bin/ld: d/typesem.o: in function `merge(Type*)':
/build/gcc-trunk/build.9/gcc/../../gcc/d/dmd/typesem.d:2029: undefined reference to `_D3dmd4root11stringtable34__T11StringValueTC3dmd5mtype4TypeZ11StringValue8toDcharsMxFNaNbNiNjZPxa'
collect2: error: ld returned 1 exit status
make: *** [../../gcc/d/Make-lang.in:236: d21] Error 1
```

**Comment by ibuclaw** [comment]
1. Don't move huge swaths of functions and their dependencies around. Move them one at a time so these can be easily bisected and reverted as-needed.
2. Don't add new global imports, use selective imports at function level.

**Comment by RazvanN7** [comment]
If we are going to ever pull out semantic routines from AST nodes then most likely this will require some collaboration from your side. Simply reverting this and posting the linker errors that have nothing to do with the functions that are being moved is not enabling me to fix this in any way.

I don't see how moving some extern(D) functions affects gdc compilation.

**Comment by ibuclaw** [comment]
> If we are going to ever pull out semantic routines from AST nodes then most likely this will require some collaboration from your side. Simply reverting this and posting the linker errors that have nothing to do with the functions that are being moved is not enabling me to fix this in any way.
> 
> I don't see how moving some extern(D) functions affects gdc compilation. 

You're doing more than moving functions though, you are changing the module dependencies.

**Comment by RazvanN7** [comment]
> You're doing more than moving functions though, you are changing the module dependencies.

Indeed, that is expected when one function is moved from one module to another. Some modules used to depend indirectly on other modules and after the refactoring the dependency becomes more clear.

**Comment by ibuclaw** [comment]
> > You're doing more than moving functions though, you are changing the module dependencies.
> 
> Indeed, that is expected when one function is moved from one module to another. Some modules used to depend indirectly on other modules and after the refactoring the dependency becomes more clear.

My earlier comment still stands though.

https://github.com/dlang/dmd/pull/15853#issuecomment-1823891157

I'd err on reverting, then move these individually, any change of imports should be kept function local.

These weird template errors arise from changing the global module dependencies.

File: compiler/src/dmd/canthrow.d [hunk 1]
```diff
@@ -22,7 +22,6 @@ import dmd.declaration;
 import dmd.dsymbol;
 import dmd.errorsink;
 import dmd.expression;
-import dmd.expressionsem;
```
Comment (RazvanN7) [inline compiler/src/dmd/canthrow.d:25]: This seems to be the only import that is added in the PR that introduced the regression. You could just put this as a local (potentially, selective) import at line 83.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Suggested fix.

**Comment by ibuclaw** [comment]
Blocked for a week, and there's [yet another pr](https://github.com/dlang/dmd/pull/15870) that pushes the diff with upstream over the 400kb limit.

**Comment by RazvanN7** [comment]
https://github.com/dlang/dmd/pull/15873 - attempt at fixing this regression.

**Comment by ibuclaw** [comment]
> @ibuclaw ok to close this?

Mainline is green again.

https://github.com/D-Programming-GDC/gcc/actions/workflows/main.yml

**Pull request closed without merging**